File: src/main/java/org/styl/gravitus/SpecsLoader.java
package org.styl.gravitus;

import java.io.File;
import java.io.IOException;

import org.apache.commons.io.FileUtils;
import org.apache.log4j.Logger;

public class SpecsLoader {
	final static Logger logger = Logger.getLogger(SpecsLoader.class);

	public static final String USER_FILE = System.getProperty("user.home") + File.separator + "Documents" +
			File.separator + "Gravitus" + File.separator + "application.properties";
	public static final String DEFAULT_FILE = "resources" + File.separator + "specs.properties";
	
	public SpecsLoader() {
		logger.info("getting properties from file");

		Specs.createInstance();
		
		File userFile = new File(USER_FILE);
		
		if(userFile.exists()) {
			try {
				Specs.instance.loadProperties(userFile);
			} catch (IOException e) {
				logger.error("application.properties file is corrupted.");
			}
		}else {
			logger.info("could not find user properties!");
			logger.info("attempting to load default properties!");
			try {
				Specs.instance.loadProperties(new File(DEFAULT_FILE));
			} catch (IOException e) {
				logger.error("failed to load properties! The application will now terminate.");
				System.exit(0);
			}
			try {
				createUserFile();
				logger.info("created user preferences.");
			} catch (IOException e) {
				logger.error("failed to create user preferences.");
			}
		}
		
	}

	private void createUserFile() throws IOException {
		FileUtils.copyFile(new File(DEFAULT_FILE), new File(USER_FILE));
	}
}

File: src/main/java/org/styl/gravitus/SpecsLoader.kt
package org.styl.gravitus

import org.apache.commons.io.FileUtils
import org.apache.log4j.Logger
import java.io.File
import java.io.IOException

class SpecsLoader {

	private val logger: Logger = Logger.getLogger(SpecsLoader::class.java)

	init {
		logger.info("getting properties from file")

		Specs.createInstance()

		val userFile = File(SystemLocations.userPropertiesFile)

		if (userFile.exists()) {
			try {
				Specs.instance.loadProperties(userFile)
			} catch (e: IOException) {
				logger.error("application.properties file is corrupted.", e)
			}
		} else {
			logger.info("could not find user properties!")
			logger.info("attempting to load default properties!")

			try {
				Specs.instance.loadProperties(File(SystemLocations.defaultPropertiesFile))
			} catch (e: IOException) {
				logger.error("failed to load properties! The application will now terminate.", e)
				kotlin.system.exitProcess(0)
			}

			try {
				createUserFile()
				logger.info("created user preferences.")
			} catch (e: IOException) {
				logger.error("failed to create user preferences.", e)
			}
		}
	}

	@Throws(IOException::class)
	private fun createUserFile() {
		FileUtils.copyFile(
			File(SystemLocations.defaultPropertiesFile),
			File(SystemLocations.userPropertiesFile)
		)
	}
}
